x/ccv/provider/keeper/misbehaviour.go

```diff
@@ -0,0 +1,154 @@
+	// construct the trusted and conflicted light blocks
+	lightBlock1, err := headerToLightBlock(*misbehaviour.Header1)
+	if err != nil {
+		return nil, err
+	}
+	lightBlock2, err := headerToLightBlock(*misbehaviour.Header2)
+	if err != nil {
+		return nil, err
+	}
+
+	var validators []*tmtypes.Validator
+
+	// compare the signatures of the headers
+	// and return the intersection of validators who signed both
+
+	// create a map with the validators' address that signed header1
+	header1Signers := map[string]struct{}{}
+	for _, sign := range lightBlock1.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		header1Signers[sign.ValidatorAddress.String()] = struct{}{}
+	}
+
+	// iterate over the header2 signers and check if they signed header1
+	for _, sign := range lightBlock2.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		if _, ok := header1Signers[sign.ValidatorAddress.String()]; ok {
+			_, val := lightBlock1.ValidatorSet.GetByAddress(sign.ValidatorAddress)
+			validators = append(validators, val)
+		}
+	}
+
+	return validators, nil
+}
+
+// headerToLightBlock returns a CometBFT light block from the given IBC header
+func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
+	sh, err := tmtypes.SignedHeaderFromProto(h.SignedHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	vs, err := tmtypes.ValidatorSetFromProto(h.ValidatorSet)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tmtypes.LightBlock{
+		SignedHeader: sh,
+		ValidatorSet: vs,
+	}, nil
+}
+
+// CheckMisbehaviour checks that headers in the given misbehaviour forms
+// a valid light client attack and that the corresponding light client isn't expired
+func (k Keeper) CheckMisbehaviour(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) error {
+	clientState, found := k.clientKeeper.GetClientState(ctx, misbehaviour.GetClientID())
+	if !found {
+		return sdkerrors.Wrapf(ibcclienttypes.ErrClientNotFound, "cannot check misbehaviour for client with ID %s", misbehaviour.GetClientID())
+	}
+
+	clientStore := k.clientKeeper.ClientStore(ctx, misbehaviour.GetClientID())
+
+	// Check that the headers are at the same height to ensure that
+	// the misbehaviour is for a light client attack and not a time violation,
+	// see ibc-go/modules/light-clients/07-tendermint/types/misbehaviour_handle.go
+	if !misbehaviour.Header1.GetHeight().EQ(misbehaviour.Header2.GetHeight()) {
+		return sdkerrors.Wrap(ibcclienttypes.ErrInvalidMisbehaviour, "headers are not at same height")
+	}
+
+	// CheckMisbehaviourAndUpdateState verifies the misbehaviour against the trusted consensus states
+	// but does NOT update the light client state.
+	// Note that the IBC CheckMisbehaviourAndUpdateState method returns an error if the trusted consensus states are expired,
+	// see ibc-go/modules/light-clients/07-tendermint/types/misbehaviour_handle.go
+	_, err := clientState.CheckMisbehaviourAndUpdateState(ctx, k.cdc, clientStore, &misbehaviour)
```

Comment:
I might be missing something in regards to light client attacks but I was wondering the following: At the end of the day, it would seem that a lunatic light client attack consists of having > 1/ 3 validators sign 2 conflicting blocks (i.e., with different `BlockID`). In other words, we still have a case of _double voting_ here (i.e., someone provided 2 votes for different blocks). If that's true, then what is the difference in slashing those malicious validators here due to misbehaviour and not just slashing them in the double-voting case. In the general case, we do freeze the client but here we do not. So, why are we making a distinction between misbehaviour and double voting?